x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,350 @@
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.NoShardAvailableActionException;
+import org.elasticsearch.action.UnavailableShardsException;
+import org.elasticsearch.action.support.TransportActions;
 import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.transport.NetworkExceptionHelper;
+import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.common.xcontent.ConstructingObjectParser;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.translog.Translog;
 import org.elasticsearch.persistent.AllocatedPersistentTask;
 import org.elasticsearch.tasks.Task;
 import org.elasticsearch.tasks.TaskId;
 
 import java.io.IOException;
+import java.util.Arrays;
+import java.util.LinkedList;
 import java.util.Map;
 import java.util.Objects;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.LongConsumer;
 
-public class ShardFollowNodeTask extends AllocatedPersistentTask {
+/**
+ * The node task that fetch the write operations from a leader shard and
+ * persists these ops in the follower shard.
+ */
+public abstract class ShardFollowNodeTask extends AllocatedPersistentTask {
+
+    static final int DEFAULT_MAX_READ_SIZE = 1024;
+    static final int DEFAULT_MAX_WRITE_SIZE = 1024;
+    static final int RETRY_LIMIT = 10;
+    static final int DEFAULT_MAX_CONCURRENT_READS = 1;
+    static final int DEFAULT_MAX_CONCURRENT_WRITES = 1;
+    static final int DEFAULT_MAX_BUFFER_SIZE = 10240;
+    static final long DEFAULT_MAX_OPERATIONS_SIZE_IN_BYTES = Long.MAX_VALUE;
+    private static final TimeValue RETRY_TIMEOUT = TimeValue.timeValueMillis(500);
+
+    private static final Logger LOGGER = Loggers.getLogger(ShardFollowNodeTask.class);
+
+    private final ShardFollowTask params;
+    private final BiConsumer<TimeValue, Runnable> scheduler;
 
-    private final AtomicLong processedGlobalCheckpoint = new AtomicLong();
+    private volatile long lastRequestedSeqno;
+    private volatile long globalCheckpoint;
 
-    ShardFollowNodeTask(long id, String type, String action, String description, TaskId parentTask, Map<String, String> headers) {
+    private volatile int numConcurrentReads = 0;
+    private volatile int numConcurrentWrites = 0;
+    private volatile long processedGlobalCheckpoint = 0;
+    private volatile long currentIndexMetadataVersion = 0;
+    private final AtomicInteger retryCounter = new AtomicInteger(0);
+    private final Queue<Translog.Operation> buffer = new LinkedList<>();
```

Comment:
I still wonder if we should use a priority queue here (ordered by sequence number) so that operations are applied closer to in order than they otherwise would be? Something like:

```java
    private final Queue<Translog.Operation> buffer = new PriorityQueue<>(Comparator.comparing(Translog.Operation::seqNo).reversed());
```